server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -1717,11 +1746,59 @@ public int cleanupJobs(AsyncJobStatusQueryBuilder queryBuilder) {
 
         queryBuilder.setJobStates(states);
 
-        // TODO: any other sanity restrictions deemed necessary
+        // Add any other sanity restrictions deemed necessary here
 
         return this.jobCurator.deleteJobs(queryBuilder);
     }
 
+    /**
+     * Aborts all non-terminal jobs matching the query parameters provided. If the query does not
+     * specify any states, this method defaults to all non-terminal, non-running states. If any
+     * terminal or running states are provided, they will be ignored.
+     *
+     * @param queryBuilder
+     *  an AsyncJobStatusQueryBuilder instance containing the various arguments or filters to use
+     *  to select jobs to abort
+     *
+     * @return
+     *  the number of jobs aborted as a result of this operation
+     */
+    @Transactional
+    public int abortNonTerminalJobs(AsyncJobStatusQueryBuilder queryBuilder) {
+        // Prepare for the defaults...
+        if (queryBuilder == null) {
+            queryBuilder = new AsyncJobStatusQueryBuilder();
+        }
+
+        Collection<JobState> states = queryBuilder.getJobStates();
+        if (states != null && !states.isEmpty()) {
+            // Make sure we don't attempt to abort some jobs that are already in terminal states
+            states = states.stream()
+                .filter(state -> state != null && !state.isTerminal() && state != JobState.RUNNING)
+                .collect(Collectors.toSet());
+        }
+        else {
+            // Set the default: all terminal states
+            states = Arrays.stream(JobState.values())
+                .filter(state -> state != null && !state.isTerminal() && state != JobState.RUNNING)
+                .collect(Collectors.toSet());
+        }
+
+        queryBuilder.setJobStates(states);
+
+        // Add any other sanity restrictions deemed necessary here
+
+        List<AsyncJobStatus> jobs = this.jobCurator.findJobs(queryBuilder);
+
+        // Step through each job and update the job state
+        for (AsyncJobStatus job : jobs) {
+            this.setJobState(job, JobState.ABORTED);
+            this.jobCurator.merge(job);
+        }
```

Comment:
This could be a performance hit. Why do we need to fetch the actual objects before changing their state?
If the only reason is to utilize the transition validation the `setJobState` does, then maybe we should be doing that validation without fetching the job objects into memory (we have the collection of JobState, we can just validate against JobState.ABORTED).

Reply:
The original intent was:
- Validate the state transition (since this is a public utility method that may get exposed via endpoint soon)
- Update the job result with some kind of abort message
- The number of jobs we're terminating this way should never be large

I ended up not doing the second one there, but I don't remember if I made a decision not to for some reason, or just plain forgot to do it. But, an update query would not prevent this behavior anyway.

An update query would, without question, be faster and more efficient here, but at the expense of the transition validation.

Reply:
I trailed off a bit, but what are you thoughts there?

Reply:
I agree that the number of jobs this will target _should_ never be large, but with the current situation in Hosted, I'd expect this to be fetching a large amount at least the first couple of times. Also, I don't think the transition validation would be an expense. We can still do that without fetching the jobs, since we already have the static list of states we're fetching for, and if the validation passes for all states, then we do the update query.

Reply:
Oof... I hadn't even consider that. I had myself stuck in the idea that every job needed state transition validation individually. I'll work on a query for bulk updating a job's state.

Reply:
Query added